### PR TITLE
Fix tmp to fn

### PR DESCRIPTION
**Pull request by cosunae**

### CODE REVIEW

**Comment by cosunae** [comment]
launch jenkins

**Comment by cosunae** [comment]
> a question on a general level:
> 
> The cuda-codegen is pretty complicated right now. Should we isolate all the caching-codegen into it's separate TU to make the path without caching easier to follow?

Yes, I thought about that. The situation is the following, we started with a PassSetCaches that determines few properties of the kcache of a field like the policy, and the window for certain policies. The specification was based on what GT needed. The problem is that our algorithms are too general, and we know now that many of the caches are not working. The cuda backend is capable of dealing with any cache and policy in a general way. However the cache specification of the PassSetCaches is not enough, and that is why we have lot of algorithmic content in the cuda code generation. 
Now that we now what we need, I would move all this algorithmic content into the pass, and keep the code generation rather simple. But I would do that in a separate work

**Comment by twicki** [comment]
I think we can have two approaches, both are fine:
- do all the algorithmic finesse in the pass, store that information. If it is passed to a backend that can handle it (aka cuda), fine, everything is there.  If it is passed to a more narrow backend (GT), we should be able to recover the relevant information that GT uses and make the codegen work with the limitations (ie full fills / flushes)
- we do the basic calculation in the pass and let every backed do additional algorithmic stuff with the depth it can handle.

I would prefer option 1, if we're not destroying information that can't be recovered?
If we go with option two though (as it is in this pr) I would suggest a separate TU ind CodeGen/Cuda/ "GenerateCaches" such that if a new developer looks at the code, one can ignore the caching code and get an easier entry into  the critical stuff (for code-legality) as this is a pure optimization-technique that could in theory be turned off. I think it would make the separation more clear. This is what I wanted to say initially

File: src/dawn/Compiler/DawnCompiler.cpp [hunk 1]
```diff
@@ -153,7 +140,7 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
   PassManager& passManager = optimizer->getPassManager();
 
   // Setup pass interface
-  optimizer->checkAndPushBack<PassInlining>(inlineStrategy);
+  optimizer->checkAndPushBack<PassInlining>(PassInlining::IK_InlineProcedures);
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/Compiler/DawnCompiler.cpp]: Why do you think so? 
isnt 
```
optimizer->checkAndPushBack<PassInlining>(PassInlining::IK_InlineProcedures);
```
more readable than
```
optimizer->checkAndPushBack<PassInlining>();
```
?

File: src/dawn/Compiler/DawnCompiler.cpp [hunk 2]
```diff
@@ -168,6 +155,9 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
   optimizer->checkAndPushBack<PassStencilSplitter>(maxFields);
   optimizer->checkAndPushBack<PassTemporaryType>();
   optimizer->checkAndPushBack<PassTemporaryMerger>();
+  if(getOptions().InlineSF || getOptions().PassTmpToFunction) {
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/Compiler/DawnCompiler.cpp]: I agree that the configuration of the pass should just be forwarded to the pass, but 
my thought here was that pushing a pass that is not executed is confusing. But I am not sure I understood your point, let talk

File: src/dawn/Optimizer/StatementMapper.cpp [hunk 3]
```diff
@@ -216,7 +216,7 @@ void StatementMapper::visit(const std::shared_ptr<StencilFunCallExpr>& expr) {
           diag << "no viable Do-Method overload for stencil function call '" << expr->getCallee()
                << "'";
           instantiation_->getOptimizerContext()->getDiagnostics().report(diag);
-          return;
+          dawn_unreachable_internal("no viable do-method overload for stencil function call");
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/Optimizer/StatementMapper.cpp]: done

File: src/dawn/IIR/DependencyGraph.h [hunk 4]
```diff
@@ -20,9 +20,11 @@
 #include <algorithm>
 #include <fstream>
 #include <iosfwd>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/IIR/DependencyGraph.h]: done

[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/DependencyGraph.h]: not yet...

File: src/dawn/Compiler/DawnCompiler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Compiler/DawnCompiler.cpp]: we came to an agreement that we pass a true/false as an argument to the pass to simplify

**Comment by twicki** [comment]
For documentation:
- We came to the conclusion that Option 1 is the desired one. The option for a more simplistic backend language is to just ignore caches that it cannot handle.
- To not clutter this up too much, we've decided to handle this in a separte PR. The corresponding issue is here: https://github.com/MeteoSwiss-APN/dawn/issues/167

File: src/dawn/Compiler/DawnCompiler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/Compiler/DawnCompiler.cpp]: done

File: src/dawn/IIR/DependencyGraph.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/IIR/DependencyGraph.h]: Yes, I forgot to push the commit, now it should be in htere

**Pull request merged**